File: dfs/组合.java
class Solution {
    List<List<Integer>> ans = new LinkedList();

    public List<List<Integer>> combine(int n, int k) {
        LinkedList<Integer> path = new LinkedList();
        // 搜索起点不一定需要枚举到 n
        for (int i = 1; i <= n - k + 1; i++) dfs(i, path, n, k);
        return ans;
    }

    private void dfs(int i, LinkedList<Integer> path, int n, int k) {
        path.add(i);
        if (path.size() == k) {
            ans.add(new LinkedList(path));
            path.removeLast();
            return;
        }

        for (int j = i + 1; j <= n; j++) dfs(j, path, n, k);
        path.removeLast();
    }
}

File: algorithm/search/code/组合.cpp
class Solution {
public:
// https://leetcode.cn/problems/combinations/
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> ans;

        function<void(vector<int>, int)> dfs = [&](vector<int> res, int index) {
            res.push_back(index);
            if (res.size() == k) {
                ans.push_back(res);
                return;
            }
            for (int i = index + 1; i <= n; i++) {
                dfs(res, i);
            }
            res.pop_back();
        };

        for (int i = 1; i <= n; i++) {
            dfs(vector<int>(0), i);
        }

        return ans;
    }

    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> ans;
        function<void(vector<int>, int)> dfs = [&](vector<int> res, int index) {
            if (index > n + 1) {
                return;
            }
            if (res.size() == k) {
                ans.push_back(res);
                return;
            }
            res.push_back(index);
            dfs(res, index + 1);
            res.pop_back();
            dfs(res, index + 1);
        };
        dfs(vector<int>(0), 1);
        return ans;
    }
};

